src/openms/source/FORMAT/CachedMzML.cpp

```diff
@@ -245,80 +269,142 @@ namespace OpenMS
     MzMLFile().store(out_meta, exp);
   }
 
-  void CachedmzML::readSpectrum_(Datavector& data1, Datavector& data2, std::ifstream& ifs, int& ms_level, double& rt) const
+  std::vector<OpenSwath::BinaryDataArrayPtr> CachedmzML::readSpectrumFast(std::ifstream& ifs, int& ms_level, double& rt)
   {
+    std::vector<OpenSwath::BinaryDataArrayPtr> data;
+    data.push_back(OpenSwath::BinaryDataArrayPtr(new OpenSwath::BinaryDataArray));
+    data.push_back(OpenSwath::BinaryDataArrayPtr(new OpenSwath::BinaryDataArray));
+
     Size spec_size = -1;
-    ifs.read((char*)&spec_size, sizeof(spec_size));
-    ifs.read((char*)&ms_level, sizeof(ms_level));
-    ifs.read((char*)&rt, sizeof(rt));
+    Size nr_float_arrays = -1;
+    ifs.read((char*) &spec_size, sizeof(spec_size));
+    ifs.read((char*) &nr_float_arrays, sizeof(nr_float_arrays));
+    ifs.read((char*) &ms_level, sizeof(ms_level));
+    ifs.read((char*) &rt, sizeof(rt));
+
+    if (static_cast<int>(spec_size) < 0)
+    {
+      throw Exception::ParseError(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION, 
+        "Read an invalid spectrum length, something is wrong here. Aborting.", "filestream");
+    }
 
-    data1.resize(spec_size);
-    data2.resize(spec_size);
+    data[0]->data.resize(spec_size);
+    data[1]->data.resize(spec_size);
 
     if (spec_size > 0)
     {
-      ifs.read((char*)&data1[0], spec_size * sizeof(DatumSingleton));
-      ifs.read((char*)&data2[0], spec_size * sizeof(DatumSingleton));
+      ifs.read((char*) &(data[0]->data)[0], spec_size * sizeof(DatumSingleton));
+      ifs.read((char*) &(data[1]->data)[0], spec_size * sizeof(DatumSingleton));
     }
+    if (nr_float_arrays == 0) return data;
+
+    char* buffer = new(std::nothrow) char[1024];
+    for (Size k = 0; k < nr_float_arrays; k++)
+    {
+      data.push_back(OpenSwath::BinaryDataArrayPtr(new OpenSwath::BinaryDataArray));
+      Size len, len_name;
+      ifs.read((char*)&len, sizeof(len));
+      ifs.read((char*)&len_name, sizeof(len_name));
+
+      // We will not read data longer than 1024 length as this is user-generated input data
+      if (len_name > 1023) ifs.seekg(len_name * sizeof(char), ifs.cur);
+      else
+      {
+        ifs.read(buffer, len_name);
+        buffer[len_name] = '\0';
+      }
+      data.back()->data.resize(len);
+      data.back()->description = buffer;
+      ifs.read((char*)&(data.back()->data)[0], len * sizeof(DatumSingleton));
+    }
+    delete[] buffer;
+    return data;
   }
 
-  void CachedmzML::readChromatogram_(Datavector& data1, Datavector& data2, std::ifstream& ifs) const
+  std::vector<OpenSwath::BinaryDataArrayPtr> CachedmzML::readChromatogramFast(std::ifstream& ifs)
```

Comment:
this looks like it does a subset of readSpectraFast().
Can this code be re-used instead of copied?